tsdb/index/tsi1/series_file.go

```diff
@@ -0,0 +1,446 @@
+package tsi1
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/influxdata/influxdb/models"
+	"github.com/influxdata/influxdb/pkg/mmap"
+	"github.com/influxdata/influxdb/pkg/rhh"
+)
+
+// ErrSeriesOverflow is returned when too many series are added to a series writer.
+var ErrSeriesOverflow = errors.New("series overflow")
+
+// Series list field size constants.
+const SeriesIDSize = 4
+
+// Series flag constants.
+const (
+	// Marks the series as having been deleted.
+	SeriesTombstoneFlag = 0x01
+
+	// Marks the following bytes as a hash index.
+	// These bytes should be skipped by an iterator.
+	SeriesHashIndexFlag = 0x02
+)
+
+const DefaultMaxSeriesFileSize = 32 * (1 << 30) // 32GB
+
+// MaxSeriesFileHashSize is the maximum number of series in a single hash.
+const MaxSeriesFileHashSize = (1048576 * LoadFactor) / 100
+
+// SeriesFile represents the section of the index that holds series data.
+type SeriesFile struct {
+	path string
+	data []byte
+	file *os.File
+	w    *bufio.Writer
+
+	offset  uint32
+	hashMap *rhh.HashMap
+
+	// MaxSize is the maximum size of the file.
+	MaxSize int64
+}
+
+// NewSeriesFile returns a new instance of SeriesFile.
+func NewSeriesFile(path string) *SeriesFile {
+	return &SeriesFile{
+		path:    path,
+		MaxSize: DefaultMaxSeriesFileSize,
+	}
+}
+
+// Open memory maps the data file at the file's path.
+func (f *SeriesFile) Open() error {
+	// Open file handler for appending.
+	file, err := os.OpenFile(f.path, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0666)
+	if err != nil {
+		return err
+	}
+	f.file = file
+
+	// Ensure header byte exists.
+	f.offset = 0
+	if fi, err := file.Stat(); err != nil {
+		return err
+	} else if fi.Size() > 0 {
+		f.offset = uint32(fi.Size())
+	} else {
+		if _, err := f.file.Write([]byte{0}); err != nil {
+			return err
+		}
+		f.offset = 1
+	}
+
+	// Wrap file write a bufferred writer.
+	f.w = bufio.NewWriter(f.file)
+
+	// Memory map file data.
+	data, err := mmap.Map(f.path, f.MaxSize)
+	if err != nil {
+		return err
+	}
+	f.data = data
+
+	// Index all series.
+	if err := f.reindex(); err != nil {
```

Comment:
This looks like it's rebuilding the hash index each time the file is open?  This needs to be an on-disk hash index that we can just open and start using or startup times will be slow.

Reply:
I still need to add persistence for the hash map. I went with a simpler approach for now so I could test the ingest performance.

Reply:
Makes sense.